portality/lib/dataobj.py

```diff
@@ -172,184 +178,731 @@ class DataSchemaException(Exception):
+        # {"identifier" : ("bibjson.identifier", DataObj))}
+        try:
+            og(self, "_properties")
+        except:
+            self._properties = properties if properties is not None else {}
+
+        # if no subclass has set expose_data, set it
+        try:
+            og(self, "_expose_data")
+        except:
+            self._expose_data = expose_data
+
+        # restructure the object based on the struct if requried
+        if self._struct is not None and raw is not None and construct_raw:
+            self.data = construct(self.data, self._struct, self._coerce_map, silent_prune=construct_silent_prune)
+
+        # run against the old validation routine
+        # (now deprecated)
+        self.validate()
+
+        # run the object's native validation routine
+        self.custom_validate()
+
+    def __getattr__(self, name):
+        props, data_attrs = self._list_dynamic_properties()
+
+        # if the name is not in the dynamic properties, raise an attribute error
+        if name not in props and name not in data_attrs:
+            raise AttributeError('{name} is not set'.format(name=name))
+
+        # otherwise, extract the path from the properties list or the internal data
+        if name in props:
+            path, wrapper = self._properties.get(name)
         else:
-            self._struct = {}
+            path = name
+            wrapper = DataObj
+
+        # request the internal property directly (which will in-turn raise the AttributeError if necessary)
+        try:
+            return self._get_internal_property(path, wrapper)
+        except AttributeError:
+            # re-wrap the attribute error with the name, rather than the path
+            raise AttributeError('{name} is not set'.format(name=name))
+
+    def __setattr__(self, key, value):
+        # first set the attribute on any explicitly defined property
+        try:
+            att = object.__getattribute__(self, key)
+            return object.__setattr__(self, key, value)
+        except AttributeError:
+            pass
+
+        # this could be an internal attribute from the constructor, so we need to make
+        # a special case
+        if key in ["_coerce_map", "_struct", "data", "_properties", "_expose_data"]:
+            return object.__setattr__(self, key, value)
+
+        props, data_attrs = self._list_dynamic_properties()
+
+        # extract the path from the properties list or the internal data
+        path = None
+        wrapper = None
+        if key in props:
+            path, wrapper = self._properties.get(key)
+        elif key in data_attrs:
+            path = key
+            wrapper = DataObj
+
+        # try to set the property on othe internal object
+        if path is not None:
+            wasset = self._set_internal_property(path, value, wrapper)
+            if wasset:
+                return
+
+        # fall back to the default approach of allowing any attribute to be set on the object
```

Comment:
So if you try to set something that is not a dynamically defined property ("shortcut" to a path) or part of the struct, we would just allow the setting to take place?

What would be the use case for this? When I wrote the previous version the reason it didn't allow you to set a property that wasn't defined in the struct was because with these objects you don't get autocomplete, which means typos are more likely, so I wanted to immediately stop you from assigning to something that isn't going to end up in self.data .